File: reference-impl/kotlin/src/main/kotlin/adr/Effect.kt
// ── Effect descriptors (seam 01) ────────────────────────────────────────────
// Plain data. The fold RETURNS these; only the boundary PERFORMS them.

package adr

sealed interface Effect {
    data class Log(val what: String, val ticket: String, val level: Priority, val at: Long) : Effect
    data class PageOncall(val ticket: String, val at: Long) : Effect
    data class Diag(val note: String) : Effect
}

File: examples/typescript/src/effect.ts
// ── Effect descriptors (seam 01) ────────────────────────────────────────────
// Plain data. The fold RETURNS these; only the boundary PERFORMS them.

import type { Priority } from "./domain";

export type Effect =
  | { readonly kind: "Log"; readonly what: string; readonly ticket: string; readonly level: Priority; readonly at: number }
  | { readonly kind: "PageOncall"; readonly ticket: string; readonly at: number }
  | { readonly kind: "Diag"; readonly note: string };
